controllers/clusters/on_premises.go

```diff
@@ -301,6 +320,168 @@ func reconcileNodesResources(ctx context.Context, b *onPremisesBootstrap) error
 	return nil
 }
 
+func reconcileIPReservations(ctx context.Context, k8sclient client.Client) (string, error) {
+	ipReservationsList := &apicalico.IPReservationList{}
+	err := k8sclient.List(ctx, ipReservationsList, &client.ListOptions{
+		LabelSelector: labels.SelectorFromSet(map[string]string{
+			models.ControlledByLabel: models.OperatorLabel,
+		}),
+	})
+	if err != nil {
+		return "", err
+	}
+
+	for _, reservation := range ipReservationsList.Items {
+		return reservation.Spec.ReservedCIDRs[0], nil
+	}
+
+	icIPPoolList := &apicalico.IPPoolList{}
+	err = k8sclient.List(ctx, icIPPoolList, &client.ListOptions{
+		LabelSelector: labels.SelectorFromSet(map[string]string{
+			models.ControlledByLabel: models.OperatorLabel,
+		}),
+	})
+	if err != nil {
+		return "", err
+	}
+
+	if len(icIPPoolList.Items) == 0 {
+		clusterIPPoolList := &apicalico.IPPoolList{}
+		err = k8sclient.List(ctx, clusterIPPoolList)
+		if err != nil {
+			return "", err
+		}
+
+		usedCIDRs := []string{}
+		for _, pool := range clusterIPPoolList.Items {
+			usedCIDRs = append(usedCIDRs, pool.Spec.CIDR)
+		}
+
+		newIPPool := apicalico.NewIPPool()
+		newIPPool.ObjectMeta.Name = models.IPPoolName
+		newIPPool.ObjectMeta.Labels = map[string]string{
+			models.ControlledByLabel: models.OperatorLabel,
+		}
+		availCIDR, err := getAvailablePrivateCIDR(usedCIDRs)
+		if err != nil {
+			return "", err
+		}
+
+		newIPPool.Spec.CIDR = availCIDR
+
+		err = k8sclient.Create(ctx, newIPPool)
+		if err != nil {
+			return "", err
+		}
+
+		icIPPoolList.Items = append(icIPPoolList.Items, *newIPPool)
+	}
+
+	newReservation := &apicalico.IPReservation{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       apicalico.KindIPReservation,
+			APIVersion: apicalico.VersionCurrent,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: models.IPReservationName,
+			Labels: map[string]string{
+				models.ControlledByLabel: models.OperatorLabel,
+			},
+		},
+		Spec: apicalico.IPReservationSpec{
+			ReservedCIDRs: []string{icIPPoolList.Items[0].Spec.CIDR},
+		},
+	}
+
+	err = k8sclient.Create(ctx, newReservation)
+	if err != nil {
+		return "", err
+	}
+
+	return newReservation.Spec.ReservedCIDRs[0], nil
+}
+
+// Calculates available CIDR from 10.*.*.* range with mask /24
+func getAvailablePrivateCIDR(unavailCIDRs []string) (string, error) {
+	for i := 0; i < 255; i++ {
+		checkNet := net.IPNet{
+			IP:   net.ParseIP(fmt.Sprintf("10.%d.0.0", i)),
```

Comment:
Why do you use the second octet and mask 24? maybe we should use the third one here?

Reply:
Yeah, added third octet